ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,525 @@
+
+NPI_MODULEID DECLSPEC_SELECTANY _ebpf_netevent_event_program_info_provider_moduleid = {
+    sizeof(NPI_MODULEID), MIT_GUID, {0}};
+
+static ebpf_extension_program_info_provider_t* _ebpf_netevent_event_program_info_provider_context = NULL;
+
+//
+// Event Hook NPI Client Attach and Detach Callbacks (to NetEvent NPI provider).
+// Callbacks invoked when a Program Information NPI client attaches/detaches.
+// (must register a separate Hook NPI provider module for each supported attach type)
+//
+static ebpf_result_t
+_netevent_ebpf_extension_netevent_event_on_client_attach(
+    _In_ const ebpf_extension_hook_client_t* attaching_client,
+    _In_ const ebpf_extension_hook_provider_t* provider_context)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    bool push_lock_acquired = false;
+
+    EBPF_EXT_LOG_ENTRY();
+
+    UNREFERENCED_PARAMETER(attaching_client);
+    UNREFERENCED_PARAMETER(provider_context);
+
+    ExAcquirePushLockExclusive(&_ebpf_netevent_event_hook_provider_lock);
+
+    push_lock_acquired = true;
+
+    if (!_ebpf_netevent_event_hook_provider_registered) {
+        // Register and attach the neteventebpfext extension to NetEvent as an NMR Client.
+        // This will invoke the _netevent_ebpf_extension_attach_provider() callback.
+        NTSTATUS status = NmrRegisterClient(
+            &_netevent_client_characteristics,
+            &_netevent_client_registration_context,
+            &_netevent_client_binding_context.nmr_binding_handle);
+        if (!NT_SUCCESS(status)) {
+            EBPF_EXT_LOG_MESSAGE_NTSTATUS(
+                EBPF_EXT_TRACELOG_LEVEL_ERROR, EBPF_EXT_TRACELOG_KEYWORD_NETEVENT, "Attach to netevent failed", status);
+            result = EBPF_OPERATION_NOT_SUPPORTED;
+            goto Exit;
+        }
+        _ebpf_netevent_event_hook_provider_registered = TRUE;
+    }
+
+    _ebpf_netevent_event_hook_provider_registration_count++;
+
+Exit:
+    if (push_lock_acquired) {
+        ExReleasePushLockExclusive(&_ebpf_netevent_event_hook_provider_lock);
+    }
+
+    EBPF_EXT_RETURN_RESULT(result);
+}
+
+static void
+_netevent_ebpf_extension_netevent_event_on_client_detach(_In_ const ebpf_extension_hook_client_t* detaching_client)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+
+    EBPF_EXT_LOG_ENTRY();
+
+    UNREFERENCED_PARAMETER(detaching_client);
+
+    // Unregister the netevent create notify routine.
+    ExAcquirePushLockExclusive(&_ebpf_netevent_event_hook_provider_lock);
+
+    _ebpf_netevent_event_hook_provider_registration_count--;
+
+    if (_ebpf_netevent_event_hook_provider_registered && _ebpf_netevent_event_hook_provider_registration_count == 0) {
+        // Detach the neteventebpfext extension from NetEvent as an NMR Client.
+        // This will invoke the _netevent_ebpf_extension_detach_provider() callback.
+        NTSTATUS status = NmrDeregisterClient(_netevent_client_binding_context.nmr_binding_handle);
```

Comment:
We're calling `NmrDeregisterClient` (An API whose behavior is beyond our control) while holding a lock.  Was curious if this was deliberate and/or necessary?  Was thinking about possible deadlocks in callback code paths.

Reply:
The NMR API is a fundamental OS component, extensively used across various OS modules, therefore very reliable.

Following the same pattern employed in `_ntos_ebpf_extension_process_on_client_detach`, the lock is utilized for the provider reference counter. When this counter reaches 0, the client will be deregistered.